### PR TITLE
[litmus] Dynamic memory allocation for kvm (and presi) modes.

### PR DESCRIPTION
**Pull request by maranget**
The technique significantly reduces executable size. This should be a real bonus for option `-driver C`.

As a consequence default memory allocation is `static` for the kvm and presi mode with option `-driver shell` and dynamic in all other configurations. Notice that presi/kvm statistics are collected only when static memory allocation is selected.

The PR also fixes a few minor problems: dead code removal, fixing printf formats not supported in kvm mode, and runtime trimming of number of instances to declared available cores.

### CODE REVIEW

**Comment by maranget** [comment]
Hi @relokin, this PR complements PR #263. With dynamic memory allocation executables are smaller. This should benefit  executables that perform several tests (option `-driver C`). As a final result, we may be able to run the complete test base as one executable (I have not tried yet).

**Comment by relokin** [comment]
Thanks Luc! This seems very useful.

I'll try it locally and have a better look at the code. I will be on holiday next week so please bare with me or merge as-is and I'll have a look when I'm back.

**Review by relokin — COMMENTED** [review]
Some comments on the code, I will try running some experiments later today too.

File: litmus/libdir/_instance.c [hunk 1]
```diff
@@ -92,39 +93,49 @@ typedef struct global_t {
   stats_t stats ;
 } global_t ;
 
+#ifndef DYNALLOC
 static global_t global  =
   { &param, &parse[0],
-    inst, role, group, mem,
+    inst, role, group,
+#ifdef DYNALLOC
```
[the remark this answers is not in the document]
Comment (maranget) [reply litmus/libdir/_instance.c:101]: Nice catch. Fixing.

File: litmus/libdir/_instance.c [hunk 2]
```diff
@@ -92,39 +93,49 @@ typedef struct global_t {
...
     0,
     SIZE_OF_TEST, NUMBER_OF_RUN, NEXE, NOCCS,
     0,
   };
+#endif
 
-static void init_global(global_t *g,int id) {
-  if (id == 0) {
+static void init_global(global_t *g) {
+#ifdef DYNALLOC
```
[the remark this answers is not in the document]
Comment (maranget) [reply litmus/libdir/_instance.c:115]: Yes we can, except for field `mem` that is set from different sources depending on allocation modes. Hence, I follow your suggestion setting the `mem` field independently.

File: litmus/libdir/_main.c [hunk 3]
```diff
@@ -58,7 +58,7 @@ int RUN(int argc,char **argv,FILE *out) {
   glo_ptr->nruns = d.max_run;
   glo_ptr->size = d.size_of_test;
   if (glo_ptr->verbose) {
-    fprintf(stderr,"%s: n=%i, r=%i, s=%i\n",prog,glo_ptr->nexe,glo_ptr->nruns,glo_ptr->size);
+    fprintf(stderr,"%s: n=%d, r=%d, s=%d\n",prog,glo_ptr->nexe,glo_ptr->nruns,glo_ptr->size);
```
[the remark this answers is not in the document]
Comment (maranget) [reply litmus/libdir/_main.c:67]: It cannot hurt to patch kvm-unit-tests...

**Comment by maranget** [comment]
> Some comments on the code, I will try running some experiments later today too.

Hi @relokin. Did you have time to run the experiments?

**Comment by relokin** [comment]
> > Some comments on the code, I will try running some experiments later today too.
> 
> Hi @relokin. Did you have time to run the experiments?

Hey @maranget, just run with the new version too. LGTM!

**Review by relokin — APPROVED** [review]

**Comment by maranget** [comment]
> Hey @maranget, just run with the new version too. LGTM!
Thanks, let's merge.

**Pull request merged**